Cython/Utility/ObjectHandling.c

```diff
@@ -280,36 +280,49 @@ static CYTHON_INLINE PyObject *__Pyx_PyObject_GetItem(PyObject *obj, PyObject* k
 
 /////////////// ObjectGetItem ///////////////
 // //@requires: GetItemInt - added in IndexNode as it uses templating.
+//@requires: ObjectHandling.c::PyObjectGetAttrStr
+//@requires: PyObjectCallOneArg
 
 #if CYTHON_USE_TYPE_SLOTS
 static PyObject *__Pyx_PyObject_GetIndex(PyObject *obj, PyObject* index) {
+    // Get element from sequence object `obj` at index `index`.
     PyObject *runerr;
     Py_ssize_t key_value;
-    PySequenceMethods *m = Py_TYPE(obj)->tp_as_sequence;
-    if (unlikely(!(m && m->sq_item))) {
-        PyErr_Format(PyExc_TypeError, "'%.200s' object is not subscriptable", Py_TYPE(obj)->tp_name);
-        return NULL;
-    }
-
     key_value = __Pyx_PyIndex_AsSsize_t(index);
     if (likely(key_value != -1 || !(runerr = PyErr_Occurred()))) {
         return __Pyx_GetItemInt_Fast(obj, key_value, 0, 1, 1);
     }
-
-    // Error handling code -- only manage OverflowError differently.
     if (PyErr_GivenExceptionMatches(runerr, PyExc_OverflowError)) {
         PyErr_Clear();
         PyErr_Format(PyExc_IndexError, "cannot fit '%.200s' into an index-sized integer", Py_TYPE(index)->tp_name);
     }
-    return NULL;
+    return NULL; 
 }
-
 static PyObject *__Pyx_PyObject_GetItem(PyObject *obj, PyObject* key) {
-    PyMappingMethods *m = Py_TYPE(obj)->tp_as_mapping;
-    if (likely(m && m->mp_subscript)) {
-        return m->mp_subscript(obj, key);
+    PyTypeObject* tp = Py_TYPE(obj);
+    PyMappingMethods *mm = tp->tp_as_mapping;
+    if (likely(mm && mm->mp_subscript)) {
+        return mm->mp_subscript(obj, key);
+    }
+
+    PySequenceMethods *sm = tp->tp_as_sequence;
+    if (likely(sm && sm->sq_item)) {
+        return __Pyx_PyObject_GetIndex(obj, key);
+    }
+    
+#if PY_VERSION_HEX >= 0x03070000
+    if (likely(PyType_Check(obj))) {
```

Comment:
I think it's ok to add the first two slot tests to the fast path function, but this third case really is 100% not performance critical. I would extract this together with the exception case below into a separate function, and then just call that here.

Basically, the fast-path is:
- go through the `mp_subscript` slot directly, if available, since we already have it in our hands.
- jump to a fast utility function if `sq_item` is available
- fall back to some generic function that handles the remaining cases.

A C compiler would turn all of this into a fairly short sequence of CPU instructions, and the `likely()` branch hints will allow it to move the final slow path far out of the way, if it considers that helpful.